### PR TITLE
Add Challonge and tournaments management features

### PR DESCRIPTION
**Pull request by papey**
This PR adds [Challonge](https://challonge.com) support to Trostani.

Here is a list of features/requirements related to this PR

- [X] Create tournament
- [x] List tournaments
- [x] Register participant
- [x] Get tournament status and results
- [x] Use a dedicated channel for some tournaments commands
- [x] Roll matchups (_new command_)
- [x] Start a tournament (_new command_)
- [x] Finish tournament (_new command_)
- [x] Use display name to identify user (_message.member.displayname_)
- [x] Add some data (ie archetype) in deckname (on ManaStack) and and append it to username on Challonge
- [x] List tournament decks (_new command_)
- [x] Report match results (_fix report bug_)

### CODE REVIEW

**Comment by papey** [comment]
Using feedbacks from tournament https://challonge.com/fr/2239e, here is a list of enhancements and bugfixes :

- [ ] status try to use current channel and check if it's a tournament one, if not, use first argument
- [ ] when reporting, ensure an uppercase on the first argument (match identifier)
- [ ] better error messages (when tournament is not opened anymore, when a match is not found)
- [ ] bugfix Discord mention parsing (unit test case with `<@!213720708402905090>`)

File: docs/docs.adoc [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+```text
+$tnmt join <archetype> (optional)
+```
+
+Example :
+
+```text
+$tnmt join Izzet Drake
+Deck
+4 Falaises des eaux vives (M20) 252
```
Comment (Namarand) [inline docs/docs.adoc:92]: This should use english card name

File: docs/docs.adoc [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+2 Island (ANA) 57
+1 Mountain (ANA) 59
+2 Forest (ANA) 60
+3 The Great Henge (ELD) 161
+1 Castle Embereth (ELD) 239
+1 Castle Garenbrig (ELD) 240
+1 Castle Vantress (ELD) 242
+3 Fabled Passage (ELD) 244
+
+Réserve
```
Comment (Namarand) [inline docs/docs.adoc:238]: Sideboard

File: src/commands/sync.ts [hunk 3]
```diff
@@ -229,6 +212,7 @@ let pushExample: string = `Temur Elementals :
 1 Castle Vantress (ELD) 242
 3 Fabled Passage (ELD) 244
 
+Réserve
```
Comment (Namarand) [inline src/commands/sync.ts:215]: Sideboard

File: src/scry/mtg.ts [hunk 4]
```diff
@@ -193,9 +193,10 @@ export class Deck {
       }
     }
 
-    if (!this.checkCardsSum(60, this.main, ">=")) {
+    const sum = this.main.reduce<number>(this.sumer, 0);
+    if (sum != 40 && sum < 60) {
```
Comment (Namarand) [inline src/scry/mtg.ts:197]: This allow a player to submit a 40 card constructed deck. Is there a way to specify deck format, so better check can be done?
You could add a "deck content check" rule, that could verify if the deck is correct based on foramt (singleton, brawl, cosntructed, historic, forgeron, limited, etc)

File: tests/mtg.test.ts [hunk 5]
```diff
@@ -99,8 +100,7 @@ class MTGDeckTestSuite extends Deck {
   }
 }
 
-let pushData = `!push Temur Elementals // standard // BO3 // Temur FTW ! :
-Deck
+let pushData = `Deck
 4 Steam Vents (GRN) 257
```
Comment (Namarand) [inline tests/mtg.test.ts:104]: English

File: tests/mtg.test.ts [hunk 6]
```diff
@@ -132,8 +132,7 @@ Sideboard
 1 Tamiyo, Collector of Tales (WAR) 220
 3 Chandra, Awakened Inferno (M20) 127`;
 
-let pushTranslateData = `!push Bant V6 // standard // BO3 // Bant ! :
-Deck
+let pushTranslateData = `Deck
 3 Île (ELD) 257
```
Comment (Namarand) [inline tests/mtg.test.ts:136]: English

File: src/scry/mtg.ts (hunk 4 above)
Comment (papey) [reply src/scry/mtg.ts:197]: Yes, it's a planned feature !

File: tests/mtg.test.ts (hunk 6 above)
Comment (papey) [reply tests/mtg.test.ts:136]: The purpose of this test is to check if translate functions works correctly from french to english.

File: src/commands/sync.ts (hunk 3 above)
Comment (papey) [reply src/commands/sync.ts]: Fixed by https://github.com/papey/trostani/commit/70e87903bf56570f72ee4120eacb866e7f686fb0

File: docs/docs.adoc (hunk 1 above)
Comment (papey) [reply docs/docs.adoc]: Fixed by https://github.com/papey/trostani/commit/d6cfe4c7b51d10229ca62a75ef564b4b2aac264c

File: src/commands/sync.ts (hunk 3 above)
Comment (papey) [reply src/commands/sync.ts]: Fixed by https://github.com/papey/trostani/commit/d6cfe4c7b51d10229ca62a75ef564b4b2aac264c

File: docs/docs.adoc (hunk 2 above)
Comment (papey) [reply docs/docs.adoc]: Fixed by https://github.com/papey/trostani/commit/d6cfe4c7b51d10229ca62a75ef564b4b2aac264c

**Pull request merged**